pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -199,32 +212,174 @@ pub mod pallet {
 			todo!()
 		}
 
-		// TODO: not sure if we really need that. Check that out with Raphael.
 		#[pallet::call_index(6)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn start_destroy(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn start_destroy(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			Self::do_start_destroy_pool(pool_id, Some(who))
 		}
 
 		#[pallet::call_index(7)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn force_start_destroy(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn force_start_destroy(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			ensure_root(origin)?;
+
+			Self::do_start_destroy_pool(pool_id, None)
 		}
 
-		// todo: check if we really need that tx.
 		#[pallet::call_index(8)]
 		#[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
-		pub fn destroy_accounts(_origin: OriginFor<T>) -> DispatchResult {
-			todo!()
+		pub fn refund_account(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			account: AccountIdLookupOf<T>,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			let who = T::Lookup::lookup(account)?;
+
+			let pool_details = Pools::<T>::get(pool_id.clone()).ok_or(Error::<T>::PoolUnknown)?;
+
+			ensure!(pool_details.state.is_destroying(), Error::<T>::InUse);
+
+			let pool_account = pool_id.clone().into();
+
+			let mut total_collateral_issuance = Self::get_pool_collateral(&pool_account);
+
+			ensure!(!total_collateral_issuance.is_zero(), Error::<T>::ZeroAmount);
+
+			let total_issuances: Vec<(FungiblesAssetIdOf<T>, FungiblesBalanceOf<T>)> = pool_details
+				.bonded_currencies
+				.iter()
+				.map(|id| (id.clone(), T::Fungibles::total_issuance(id.clone())))
+				.filter(|(_, iss)| iss.gt(&Zero::zero()))
+				.collect();
+
+			let mut sum_of_issuances: FungiblesBalanceOf<T> = total_issuances
+				.iter()
+				.fold(Zero::zero(), |sum, (_, x)| sum.saturating_add(*x));
+
+			ensure!(!sum_of_issuances.is_zero(), Error::<T>::ZeroAmount);
+
+			let mut dead_currencies = Vec::with_capacity(total_issuances.len());
+			for (currency_id, token_issuance) in total_issuances.iter() {
+				let burnt = T::Fungibles::decrease_balance(
+					currency_id.clone(),
+					&who,
+					Bounded::max_value(),
+					Precision::BestEffort,
+					Preservation::Expendable,
+					Fortitude::Force,
+				)?;
+
+				let amount = burnt
+					.checked_mul(&total_collateral_issuance)
+					.ok_or(ArithmeticError::Overflow)? // TODO: do we need a fallback if this fails?
+					.checked_div(&sum_of_issuances)
+					.unwrap_or(Zero::zero());
+
+				T::CollateralCurrency::transfer(
+					T::CollateralAssetId::get(),
+					&pool_account,
+					&who,
+					amount,
+					Preservation::Expendable,
+				)?;
+
+				total_collateral_issuance -= amount;
+				sum_of_issuances -= burnt;
```

Comment:
these are unchecked bc the transfer amount can't be larger than the collateral, and the burnt amount can't be larger than the issuance of all currencies combined. Or should I still check?

Reply:
I would still use the saturated trait. Just for safety

Reply:
15369478fd8d5b498cdb7b59f0941bf5bee7fca8